File: backend/registration-service/app/com/radovan/play/services/impl/EurekaServiceDiscoveryImpl.java
package com.radovan.play.services.impl;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.radovan.play.services.EurekaServiceDiscovery;
import jakarta.inject.Inject;
import jakarta.inject.Singleton;
import play.libs.ws.WSClient;
import play.libs.ws.WSResponse;

import java.util.Iterator;
import java.util.Optional;

@Singleton
public class EurekaServiceDiscoveryImpl implements EurekaServiceDiscovery {

    private static final String EUREKA_API_SERVICES_URL = "http://localhost:8761/eureka/apps";
    private final ObjectMapper objectMapper = new ObjectMapper();
    private final WSClient wsClient;

    @Inject
    public EurekaServiceDiscoveryImpl(WSClient wsClient) {
        this.wsClient = wsClient;
    }

    @Override
    public String getServiceUrl(String serviceName) {
        WSResponse response = wsClient.url(EUREKA_API_SERVICES_URL)
                .addHeader("Accept", "application/json")
                .get()
                .toCompletableFuture()
                .join();

        JsonNode responseJson = response.asJson();

        if (responseJson == null || responseJson.isEmpty()) {
            throw new RuntimeException("No services found in Eureka registry");
        }

        JsonNode applicationsNode = responseJson.get("applications");
        if (applicationsNode == null) {
            throw new RuntimeException("No applications node in Eureka response");
        }

        JsonNode appsNode = applicationsNode.get("application");
        if (appsNode == null || !appsNode.isArray()) {
            throw new RuntimeException("No application array in Eureka response");
        }

        Iterator<JsonNode> apps = appsNode.elements();
        while (apps.hasNext()) {
            JsonNode app = apps.next();
            String appName = Optional.ofNullable(app.get("name"))
                    .map(JsonNode::asText)
                    .orElse(null);

            if (appName != null && appName.equalsIgnoreCase(serviceName)) {
                JsonNode instancesNode = app.get("instance");
                if (instancesNode != null && instancesNode.isArray()) {
                    Iterator<JsonNode> instances = instancesNode.elements();
                    while (instances.hasNext()) {
                        JsonNode instance = instances.next();
                        String address = Optional.ofNullable(instance.get("hostName"))
                                .map(JsonNode::asText)
                                .orElse(null);
                        Integer port = Optional.ofNullable(instance.get("port"))
                                .map(p -> p.get("$"))
                                .map(JsonNode::asInt)
                                .orElse(null);

                        if (address != null && port != null) {
                            return "http://" + address + ":" + port;
                        }
                    }
                }
            }
        }

        throw new RuntimeException("Service not found: " + serviceName);
    }
}

File: backend/infraction-service/app/com/radovan/play/services/impl/EurekaServiceDiscoveryImpl.scala
package com.radovan.play.services.impl

import com.fasterxml.jackson.databind.{JsonNode, ObjectMapper}
import com.radovan.play.services.EurekaServiceDiscovery
import jakarta.inject.{Inject, Singleton}
import play.libs.ws.WSClient

import scala.jdk.CollectionConverters._


@Singleton
class EurekaServiceDiscoveryImpl @Inject() (
                                             wsClient: WSClient
                                           ) extends EurekaServiceDiscovery {

  private val EUREKA_API_SERVICES_URL = "http://localhost:8761/eureka/apps"
  private val objectMapper = new ObjectMapper()

  override def getServiceUrl(serviceName: String): String = {
    val responseJson: JsonNode = wsClient.url(EUREKA_API_SERVICES_URL)
      .addHeader("Accept", "application/json")
      .get()
      .toCompletableFuture
      .join()
      .asJson()

    if (responseJson == null || responseJson.isEmpty)
      throw new RuntimeException("No services found in Eureka registry")


    val appsNodeOpt = Option(responseJson.get("applications")).flatMap(n => Option(n.get("application")))
    val apps = appsNodeOpt.map(_.elements().asScala).getOrElse(Seq.empty)

    for (app <- apps) {
      val appNameOpt = Option(app.get("name")).map(_.asText())
      if (appNameOpt.exists(_.equalsIgnoreCase(serviceName))) {
        val instancesOpt = Option(app.get("instance")).map(_.elements().asScala)
        for {
          instances <- instancesOpt
          instance <- instances
          address <- Option(instance.get("hostName")).map(_.asText())
          port <- Option(instance.get("port")).flatMap(p => Option(p.get("$")).map(_.asInt()))
        } {
          return s"http://$address:$port"
        }
      }
    }

    throw new RuntimeException(s"Service not found: $serviceName")
  }


}